posts/build-a-pwa-in-jhipster.md

```diff
@@ -0,0 +1,183 @@
+```html
+<script>
+    if ('serviceWorker' in navigator) {
+        window.addEventListener('load', function () {
+            navigator.serviceWorker.register('/service-worker.js')
+                .then(function () {
+                    console.log('Service Worker Registered');
+                });
+        });
+    }
+</script>
+```
+When we run the production build for our frontend Webpack will generate our service worker in the build folder for us. However, by itself, that file will not do anything. In order to add the caching and offline support we want, we must first register the service worker on our domain.
+
+
+## Installing on the Home Screen
+Now that we got through the hard parts, the last file is the `manifest.webapp`. The manifest file is metadata about your app in the form of JSON to provide details for the application installed on the home screen.
+```json
+{
+  "name": "MyPWA",
+  "short_name": "MyPWA",
+  "icons": [
+      {
+          "src": "./content/images/jhipster_family_member_0_head-192.png",
+          "sizes": "192x192",
+          "type": "image/png"
+      },
+      {
+          "src": "./content/images/jhipster_family_member_0_head-256.png",
+          "sizes": "256x256",
+          "type": "image/png"
+      },
+      {
+          "src": "./content/images/jhipster_family_member_0_head-384.png",
+          "sizes": "384x384",
+          "type": "image/png"
+      },
+      {
+          "src": "./content/images/jhipster_family_member_0_head-512.png",
+          "sizes": "512x512",
+          "type": "image/png"
+      }
+  ],
+  "theme_color": "#000000",
+  "background_color": "#000000",
+  "start_url": "/index.html",
+  "display": "standalone",
+  "orientation": "portrait"
+}
+```
+The entire file is generated for you and linked in the `index.html`. All of this specifies content on how the app will open when installed on a device, as well as, what the icon will look like to open it.
+
+In order for a user to install a progressive web app from their browser, it must meet the following criteria:
+- [X] The web app is not already installed.
+- [X] Includes a web app manifest.
+- [X] Has a registered service worker.
+- [] Served over HTTPS (required for service workers).
+- [] The user meets the engagement heuristic (The user has interacted with the domain for at least 30 seconds)
+
+We have met almost all of the criteria for installing our app, the last parts are handled by our deployment and having users engage with the app. For my PWA, I chose to host the app in S3 and use CloudFront to deliver the content over HTTPS. Be careful if you choose to serve your content through CloudFront. You must be sure to invalidate your index, the service worker, and your manifest files. Otherwise, users who have already visited your site will have old content delivered from Cloudfront, which will match the existing content in their cache.
```

Comment:
```suggestion
We have met almost all of the criteria for installing our app, the last parts are handled by our deployment and having users engage with the app. For my PWA, I chose to host the app in S3 and use CloudFront to deliver the content over HTTPS. Be careful if you choose to serve your content through CloudFront. You must be sure to invalidate your index, the service worker, and your manifest files. Otherwise, users who have already visited your site will have old content delivered from CloudFront, which will match the existing content in their cache.
```